src/MulticallerWithSigner.sol

```diff
@@ -0,0 +1,565 @@
+            // This is the also the byte length of the `targets` array and `values` array.
+            data.length := shl(5, data.length)
+
+            /* -------------------- CHECK SIGNATURE --------------------- */
+
+            // Layout the fields of the struct hash.
+            mstore(returndatasize(), AGGREGATE_WITH_SIGNER_TYPEHASH)
+            // Compute and store `keccak256(abi.encodePacked(message))`.
+            mstore(0x20, keccak256(add(message, 0x20), mload(message)))
+            // Compute and store `keccak256(abi.encodePacked(targets))`.
+            calldatacopy(0x40, targets.offset, data.length)
+            mstore(0x40, keccak256(0x40, data.length))
+            // Compute and store `keccak256(abi.encodePacked(keccak256(data[0]), ..))`.
+            for { let i := returndatasize() } iszero(eq(i, data.length)) { i := add(i, 0x20) } {
+                let o := add(data.offset, calldataload(add(data.offset, i)))
+                let p := add(0x60, i)
+                calldatacopy(p, add(o, 0x20), calldataload(o))
+                mstore(p, keccak256(p, calldataload(o)))
+            }
+            mstore(0x60, keccak256(0x60, data.length))
+            // Compute and store `keccak256(abi.encodePacked(values))`.
+            calldatacopy(0x80, values.offset, data.length)
+            mstore(0x80, keccak256(0x80, data.length))
+            mstore(0xa0, nonce) // Store the nonce.
+            mstore(0xc0, sload(or(shl(96, signer), 1))) // Store the nonce salt.
+            mstore(0x40, keccak256(returndatasize(), 0xe0)) // Compute and store the struct hash.
+            // Layout the fields of the domain separator.
+            mstore(0x80, _DOMAIN_TYPEHASH)
+            mstore(0xa0, _NAME_HASH)
+            mstore(0xc0, _VERSION_HASH)
+            mstore(0xe0, chainid())
+            mstore(0x100, address())
+            mstore(0x20, keccak256(0x80, 0xa0)) // Compute and store the domain separator.
+            // Layout the fields of `ecrecover`.
+            mstore(returndatasize(), 0x1901) // Store "\x19\x01".
+            mstore(returndatasize(), keccak256(0x1e, 0x42)) // Compute and store the digest.
+            mstore(0x20, byte(returndatasize(), calldataload(add(signature.offset, 0x40)))) // `v`.
+            calldatacopy(0x40, signature.offset, 0x40) // Copy `r` and `s`.
+            pop(
+                staticcall(
+                    gas(), // Amount of gas left for the transaction.
+                    and(eq(signature.length, 65), lt(mload(0x60), _S_THRES)), // `ecrecover`.
+                    returndatasize(), // Start of input.
+                    0x80, // Size of input.
+                    returndatasize(), // Start of output.
+                    0x20 // Size of output.
+                )
+            )
+            // `returndatasize()` will be 32 upon success and 0 otherwise.
+            let recoverySuccess := mul(returndatasize(), eq(mload(0x00), signer))
```

Comment:
Should we be checking for `signer == address(0)` or is this handled somewhere? Could lead to authorization weirdness, e.g. if some token implements this.

Reply:
We know that the high-level Solidity `ecrecover` returns the zero address if the `staticcall` fails.

But as we are in assembly, we will only need to check that the `returndatasize()` is non-zero.

You can look at the code of the clients to see that the `ecrecover` returns a truncated `keccak256` result if the `staticcall` succeeds. 

https://github.com/ethereum/go-ethereum/blob/b058cf454b3bdc7e770e2b3cec83a0bcb48f55ee/core/vm/contracts.go#L188

Reply:
The more you know...

```yul
mstore(0, 0)

let _11 := staticcall(gas(), 1 , _9, sub(_10, _9), 0, 32)
if iszero(_11) { revert_forward_1() }
let expr_23 := shift_left_0(mload(0))
/// @src 0:707:748  "address signer = ecrecover(hash, v, r, s)"
let var_signer_17 := expr_23
```